File: C++/Leetcode 472.cpp
#include <stdio.h>
#include <bits/stdc++.h>
using namespace std;

class Solution {
    unordered_map<string,int> umap;

    bool Helper(string A)
    {
        for (int i = 1; i < A.length(); i++)
        {
            string left = A.substr(0,i);
            string right = A.substr(i,A.length() - i);

            if (umap[left] && (umap[right] || Helper(right)))
            {
                return true;
            }
        }
        return false;
    }
public:
    vector<string> findAllConcatenatedWordsInADict(vector<string>& words) {
        for (int i = 0; i < words.size(); i++)
        {
            umap[words[i]]++;
        }

        vector<string> answer;
        for(int i = 0; i < words.size(); i++)
        {
            if (Helper(words[i]))
            {
                answer.push_back(words[i]);
            }
        }
        return answer;
    }
};

File: Python/Leetcode 472.py
class Solution:
    def Helper(self, word : string, temp_dict : dict) -> bool:
        for i in range(1,len(word)):
            left = word[:i]
            right = word[i:len(word)]

            if (left in temp_dict.keys()) and (right in temp_dict.keys() or self.Helper(right,temp_dict)):
                return True
        
        return False

    def findAllConcatenatedWordsInADict(self, words: List[str]) -> List[str]:
        concatenated_words = []
        temp_dict = dict() 

        for word in words:
            temp_dict[word] = 1

        for word in words:
            if self.Helper(word,temp_dict):
                concatenated_words.append(word)
        

        return concatenated_words